loader/vk_loader_platform.h

```diff
@@ -103,7 +103,7 @@ static inline bool loader_platform_is_path_absolute(const char *path) {
 
 static inline char *loader_platform_dirname(char *path) { return dirname(path); }
 
-#if defined(__linux__)
+#if defined(__linux__) || defined(__FreeBSD__)
```

Comment:
FreeBSD neither mounts `/proc` by default nor contains `/proc/self/exe`. Consider using [downstream patch](https://github.com/freebsd/freebsd-ports/blob/7caf52e37ba2174dc50a7440d63eec70d7d58ebc/graphics/vulkan-loader/files/patch-loader_vk__loader__platform.h) instead.

Reply:
I fixed the previous feedback, but I didn't get this error.

If I run vkcube I see this error instead:
```
$ ./vkcube --validate
vkEnumerateInstanceLayerProperties failed to find required validation layer.

Please look at the Getting Started guide for additional information.
```
I am debugging my setup to reproduce it, I suppose there are more problems and I don't see it... :/

Reply:
> vkEnumerateInstanceLayerProperties failed to find required validation layer

`vkcube --validate` requires https://github.com/KhronosGroup/Vulkan-ValidationLayers. Downstream it can be obtained via `pkg install vulkan-validation-layers`.

Reply:
Yes, indeed, thank you! The problem is that I hadn't modified them yet to run on FreeBSD and so I need to do this first :) (FreeBSD is not in the supported platforms in layers/vk_platform.h)

Reply:
I hadn't look at this issue for a long time sorry. So, I've rebased my branches applied the patch, modified the Validation-Layers to support FreeBSD and reproduced the issue. The problem is that even with the patch I am not able to call `vkCreateDevice` in either `vkcube` or `vkcubepp` successfully. The problem seems to be that in `anv_device_import_bo_from_host_ptr` (that's in the Linux mesa driver) the `host_ptr` is invalid. I am trying to understand why and fix it. My GPU is an Intel Ivybridge, if using the patch with another GPU works for you then I guess the problem is in the driver, but I think the problem is in my branch... so I am debugging it :)

Reply:
> anv_device_import_bo_from_host_ptr

FreeBSD doesn't support `CONFIG_MMU_NOTIFIER` yet, so userptr works only as root (via `I915_USERPTR_UNSYNCHRONIZED`). See [downstream workaround](https://github.com/freebsd/freebsd-ports/blob/main/graphics/mesa-devel/files/patch-userptr)

anv since Brodwell (Gen >= 8) uses softpin instead of userptr. However, VK_EXT_external_memory_host still needs userptr.

Reply:
Note, userptr was made optional since Mesa 19.0.0 via https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/11

Reply:
Thank you very much for this information! I am checking out the latest code of mesa to test it. Would you mind if I submit the patch (with your details I mean) to mesa on FreeDesktop Gitlab too? I've already sent some changes of mine to run crocus/iris on FreeBSD, that are now merged, and I was able to build the driver and run OpenGL applications on my IvyBridge with just the main branch. I didn't run into this problem when I tried all that, and so, I wasn't aware about the differences! Or feel free to send the patch yourself if you prefer and I can review it, it would be nice to have it on mesa! I am testing it! Thanks!

Reply:
> Would you mind if I submit the patch ... on FreeDesktop Gitlab too?

Go ahead. I only have Skylake, so can't test crocus while testing `VK_EXT_external_memory_host` and Clover (on iris) is non-trivial.

> I didn't run into this problem when I tried all that

iris/crocus probably only need userptr for [experimental Clover support](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/7047). I've never tested it due to already maintaining [compute-runtime package](https://www.freshports.org/lang/intel-compute-runtime) (requires userptr) which has better OpenCL support and Level Zero (used by ISPC and MPICH).

Reply:
Note, current version of the anv workaround was [tested on Haswell](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=244877) but actually adapted from compute-runtime tested on Skylake. Before that the patch existed as [a series of hacks](https://bugs.freebsd.org/bugzilla/attachment.cgi?id=186610) during [initial porting](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=221540) but got obsolete by [LinuxKPI fixes](https://reviews.freebsd.org/D23846) and Mesa 19.0.0 release.

Reply:
By the way, thank you for upstreaming FreeBSD support. I spend too much time downstream, so have trouble upstreaming everything.

Reply:
`Note, current version of the anv workaround was tested on Haswell after adapting from compute-runtime tested on Skylake. Before that the patch existed as a series of hacks during initial porting but got obsolete by LinuxKPI fixes and Mesa 19.0.0 release.`

It seems to work well on IvyBridge too (but as you said above for the root user only). I've tested it with Vulkan-Tools and a hobby project of mine though and none of them imports memory objects, so I can't verify VK_EXT_external_memory_host is working too. Do you use CTS to test the extensions?

Reply:
This wasn't addressed completely. Despite `KERN_PROC_PATHNAME` code `/proc/self/exe` is still used due to the order of ifdefs where `defined(__FreeBSD__)` appears.
```suggestion
#if defined(__linux__)
```

Reply:
Indeed I see this `if defined(__FreeBSD__) ... elif defined(__FreeBSD__)`, sorry about that I am going to fix it in a new PR. Removing it and including <sys/types.h> in elif gives me some compile errors though, I'll fix them and send a new patch tomorrow.

Reply:
New PR: https://github.com/KhronosGroup/Vulkan-Loader/pull/738